File: apps/api/schemas/research.py
"""
Pydantic schemas for API request/response models.
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Request Models ──────────────────────────────────────────

class ResearchRequest(BaseModel):
    """Request body for starting a new research run."""
    question: str = Field(..., min_length=5, max_length=2000, description="The research question")
    depth: str = Field("standard", pattern="^(quick|standard|deep)$")
    sources: list[str] = Field(default=["web"], description="Source types to search")
    max_sources: int = Field(20, ge=5, le=50)
    require_citation_verification: bool = True
    allow_autonomous_research: bool = True


# ── Response Models ─────────────────────────────────────────

class ResearchResponse(BaseModel):
    """Response after creating a research run."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    question: str
    status: str
    created_at: datetime


class ClaimResponse(BaseModel):
    """A single claim with its verdict and confidence."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    text: str
    verdict: str
    confidence: float
    evidence_span: Optional[str] = None
    source_id: Optional[UUID] = None
    created_at: datetime


class SourceResponse(BaseModel):
    """A discovered source with metadata."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    url: str
    title: Optional[str] = None
    author: Optional[str] = None
    source_type: Optional[str] = None
    published_at: Optional[datetime] = None
    quality_score: Optional[float] = None
    created_at: datetime


class EventResponse(BaseModel):
    """An agent execution event."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    agent_name: str
    event_type: str
    data: dict
    tokens_used: Optional[int] = None
    latency_ms: Optional[float] = None
    created_at: datetime


class ResearchDetailResponse(ResearchResponse):
    """Full research run details with claims and sources."""
    model_config = ConfigDict(from_attributes=True)

    depth: Optional[str] = None
    config: Optional[dict] = None
    result: Optional[dict] = None
    stats: Optional[dict] = None
    updated_at: Optional[datetime] = None

File: apps/web/src/types/research.ts
export type ResearchStatus = 'pending' | 'planning' | 'researching' | 'extracting' | 'verifying' | 'synthesizing' | 'completed' | 'failed';

export interface ResearchConfig {
  depth: 'quick' | 'standard' | 'deep' | string;
  sources: string[];
  max_sources: number;
  require_citation_verification: boolean;
  allow_autonomous_research: boolean;
}

export interface ResearchStats {
  duration_ms: number;
  sources_analyzed: number;
  claims_extracted: number;
  claims_verified: number;
  contradictions_found: number;
  citations_count: number;
  avg_confidence: number;
  tokens_used: number;
}

export interface ResearchRun {
  id: string;
  question: string;
  status: ResearchStatus;
  depth: string;
  config: ResearchConfig;
  result?: string | { report?: string; sources_count?: number; [key: string]: any };
  stats: ResearchStats;
  created_at: string;
  updated_at?: string;
  claims?: Claim[];
  sources?: Source[];
}

export interface Claim {
  id: string;
  text: string;
  verdict: 'SUPPORTED' | 'PARTIALLY_SUPPORTED' | 'CONTRADICTED' | 'UNVERIFIED' | string;
  confidence: number;
  evidence_span?: string;
  source_id?: string;
  source_title?: string;
  created_at?: string;
}

export interface Source {
  id: string;
  url: string;
  title?: string;
  author?: string;
  source_type?: string;
  published_at?: string;
  quality_score?: number;
  created_at?: string;
}

export interface Citation {
  id: string;
  claim_id: string;
  source_id: string;
  passage?: string;
  relevance_score?: number;
}
